alt-tab-macos/ui/PreferencesPanel.swift

```diff
@@ -1,110 +1,263 @@
     }
 
-    private func makeLabelsAndInputs() -> [[NSView]] {
+    override func close() {
+        NSApp.deactivate()
+        (NSApp as! Application).preferencesPanel = nil
+        super.close()
+    }
+
+    public func windowShouldClose(_ sender: NSWindow) -> Bool {
+        windowCloseRequested = true
+        challengeNextInvalidEditableTextField()
+        return attachedSheet == nil // user is challenged
+    }
+
+    private func challengeNextInvalidEditableTextField() {
+        let invalidFields = (contentView?
+                .findNestedViews(subclassOf: PreferencesPanelNSTextFieldEditable.self)
+                .filter({ !$0.isValid() })
+        )
+        let focusedField = invalidFields?.filter({ $0.currentEditor() != nil }).first
+        let fieldToNotify = focusedField ?? invalidFields?.first
+        fieldToNotify?.delegate?.controlTextDidChange?(Notification(name: NSControl.textDidChangeNotification, object: fieldToNotify))
+
+        if fieldToNotify != focusedField {
+            makeFirstResponder(fieldToNotify)
+        }
+    }
+
+    private func makeContentView() -> NSView {
+        let wrappingView = NSStackView(views: makePreferencesViews())
+        let contentView = NSView()
+        contentView.addSubview(wrappingView)
+
+        // visual setup
+        wrappingView.orientation = .vertical
+        wrappingView.alignment = .left
+        wrappingView.spacing = panelPadding * 0.3
+        wrappingView.topAnchor.constraint(equalTo: contentView.topAnchor, constant: panelPadding * 0.5).isActive = true
+        wrappingView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: panelPadding * -0.5).isActive = true
+        wrappingView.leftAnchor.constraint(equalTo: contentView.leftAnchor, constant: panelPadding * 0.5).isActive = true
+        wrappingView.rightAnchor.constraint(equalTo: contentView.rightAnchor, constant: panelPadding * -0.5).isActive = true
+
+        return contentView
+    }
+
+    private func makePreferencesViews() -> [NSView] {
+        // TODO: make the validators be a part of each Preference
+        let tabKeyCodeValidator: ((String)->Bool) = {
+            guard let int = Int($0) else {
+                return false
+            }
+            // non-special keys (mac & pc keyboards): https://eastmanreference.com/complete-list-of-applescript-key-codes
+            var whitelistedKeycodes: [Int] = Array(0...53)
+            whitelistedKeycodes.append(contentsOf: [65, 67, 69, 75, 76, 78, ])
+            whitelistedKeycodes.append(contentsOf: Array(81...89))
+            whitelistedKeycodes.append(contentsOf: [91, 92, 115, 116, 117, 119, 121])
+            whitelistedKeycodes.append(contentsOf: Array(123...126))
+            return whitelistedKeycodes.contains(int)
+        }
+        let maxThumbnailsPerRowValidator: ((String)->Bool) = { (3...16).contains(Int($0) ?? -1) }
+        let windowDisplayDelayValidator: ((String)->Bool) = { (0..<10000).contains(Int($0) ?? -1) }
```

Comment:
See, you go really detailed on the validation here, but you don't surface that to the user. I think my comment above about just putting sliders is the way to go. For the tab key, I think for now let's let the user type an incorrect int for now. We are gonna replace this soon with a lib, so no need to spend time here I think

Reply:
You are right, the special nature of this input is just hinted `KeyCode (48 = Tab)` wording. 

This is just an improvement pull request for the related issue and I required some kind of fast and simple validation when I added that red visualisation state to the other textfields: an invalid key code would be a valid CGFloat (so the panel would not throw any error for that). This keycodes array corrects that and is an attempt on an easy fix/improvement to the current situation.

I planned to link that nice keycodes url as a 'good enough' solution via a button or clickable text (in the grey suffix textfield) in an upcoming iteration of this PreferencesPanel since its planned to integrate/implement some keycode recording in the future.

I think Sliders and PopUps for everything are also the way to go here and TextFields only where it makes absolute sense to use one. They overcomplicate this Panel and open a can of worms with validation & visualisation.